File: generics/generics.cpp
#include <vector>
#include <string>

using std::vector;
using std::string;

int main () {
    vector<int> ints = {};
    // We can add values of the same type...
    ints.push_back(1);
    ints.push_back(2);

    // but not of others!
    ints.push_back("hello"); // Causes compilation to fail

    vector<string> strings = {"hello", "world"};

    // Add another string.
    strings.push_back("!");

    // Once again, only the type specified will be accepted!
    strings.push_back(1); // Causes compilation to fail
}

File: generics/generics.java
import java.util.ArrayList;

public class generics {

    public static void main(String[] args) {
        ArrayList<String> strings = new ArrayList<String>();
        
        // We can add values to the ArrayList...
        strings.add("Hello");
        strings.add("World");

        // But only of the declared type!
        strings.add(1); // Causes compilation to fail

        ArrayList<Integer> ints = new ArrayList<Integer>();
        ints.add(2187);
        ints.add(42);
        
        // Once again, only the declared type can be added.
        ints.add(93.2); // Causes compilation to fail
    }
}
